File: credit.c
#include <stdio.h>
#include <cs50.h>
#include <math.h>
#include <stdlib.h>
#define _GNU_SOURCE
#include <string.h>


//function prototype
long give_digit(long input);
bool legit (long card_no);
string begins_with(string cc, int x);

int main(void)
{
    string cc_string = get_string("Input: ");
    //printf("String Lenght %li\n", strlen(cc_string));
    //printf("CC begins with %s\n", begins_with(cc_string, 1));
    //printf("First 2 digit is %c%c\n", cc_string[0], cc_string[1]);
    if ( strlen(cc_string) == 15 && legit(atol(cc_string)) && strcmp(begins_with(cc_string, 2), "37") == 0)
    {
        printf("AMEX\n");
    }
   else if ( strlen(cc_string) == 16 && legit(atol(cc_string)) && strcmp(begins_with(cc_string, 1), "5") == 0)
    {
        printf("MASTERCARD\n");
    }
    else if ( strlen(cc_string) >= 13 && legit(atol(cc_string)) && strcmp(begins_with(cc_string, 1), "4") == 0)
    {
        printf("VISA\n");
    }
    else
    {
        printf("INVALID\n");
    }

}


bool legit (long card_no)
{

    //long cc = get_long("Number: ");
    int i = 1;
    long sum1 = 0;
    long sum2 = 0;
    while (card_no > 0 && i >= 1)
    {
        if (i % 2 == 0 && give_digit (card_no) * 2 >= 10)
        {
            sum1 += give_digit(give_digit (card_no) * 2) + 1;
            card_no/=10;
            i++;
        }
        else if (i % 2 == 0)
        {
            sum1 += give_digit(card_no) * 2;
            card_no/=10;
            i++;
        }
        else
        {
            sum2 += give_digit(card_no);
            card_no/=10;
            i++;
        }
    }
    int Total = sum1 + sum2;
    if ( give_digit(Total) == 0 )
    {
        return true;
    }
    return false;

}


//function that returns a single digit from the input
long give_digit(long input)
{
    long n = input % 10;
    return n;
}

//begins_with returns first 1 or 2 digit of the credit card no.
string begins_with(string cc, int x)
{
    if (x == 1)
    {
        char c[2] = {cc[0],'\0'};
        string y = c;
        return y;
    }
    else if (x == 2)
    {
        char c[3] = {cc[0], cc[1], '\0'};
        string y = c;
        return y;
    }
    else
    {
        return 0;
    }
}

File: credit.py
import re
from cs50 import get_string

sum1 = sum2 = 0
CC = get_string("Input: ")
# Put CC to x do that it becomes an integer for easier calculation
x = int(CC)

# Just like in C program in Week 1, x % 10 will spits out the end digit of the CC.
for i in range(len(CC)):
    if i % 2 == 1 and (x % 10) * 2 >= 10:
        sum1 += ((x % 10) * 2) % 10 + 1
# need to do double divide ("//") here as it will spit out a whole number of the CC,
# whereas if you just do "X/=10", it gives you decimal.
        x //= 10
    elif (i % 2 == 1):
        sum1 += (x % 10) * 2
        x //= 10
    else:
        sum2 += x % 10
        x //= 10
Total = sum1 + sum2

# Conditional checking, no dramas here, only if statements according to the CC validation rules.
if len(CC) == 15 and Total % 10 == 0 and CC[0] == '3' and CC[1] == '7':
    print("AMEX")
elif len(CC) == 16 and Total % 10 == 0 and CC[0] == '5':
    print("MASTERCARD")
elif len(CC) >= 13 and Total % 10 == 0 and CC[0] == '4':
    print("VISA")
else:
    print("INVALID")